File: 076-最小覆盖子串.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        //采用滑动窗口法
        if(t.size() > s.size())
            return "";
        string result;
        int left = 0 , count = 0, minlen = s.size() + 1;  //count记录s中出现的t中字符的个数，minlen记录最小覆盖子串长度
        unordered_map<char,int> tmap;  //用hash表记录t的字符
        for(int i = 0;i<t.size();i++)
        {
                tmap[t[i]]++;
        }
        for(int right = 0;right<s.size();right++)
        {
            if(tmap.find(s[right]) != tmap.end())  //s[i]是t中的字符
            {
                tmap[s[right]]--;
                if(tmap[s[right]] >= 0)  //统计已经出现的t的字符时防止重复出现的字符
                    count++;
                while(count == t.size())  //出现了覆盖子串
                {
                    if((right - left + 1) < minlen)  //更新当前的最小覆盖子串信息
                    {
                        minlen = right - left + 1;
                        result = s.substr(left, minlen);
                    }
                    if(tmap.find(s[left]) != tmap.end())   //向右移动窗口的左下标
                    {
                        tmap[s[left]]++;  //重新寻找覆盖子串，更新之前的tmap和count的信息
                        if(tmap[s[left]] > 0)   //tmap[s[left]] > 0说明s中还需要找覆盖的字符，count--；
                            count--;
                    }
                    left++;
                }
                
            }
        }
        return result;
    }
};

File: Java/076-最小覆盖子串.java
class Solution {
    public String minWindow(String s, String t) {
        String string = "";

        // hashmap来统计t字符串中各个字母需要出现的次数
        HashMap<Character, Integer> map = new HashMap<>();
        for (char c : t.toCharArray())
            map.put(c, map.containsKey(c) ? map.get(c) + 1 : 1);
        // 用来计数 判断当前子字符串中是否包含了t中全部字符
        int count = 0;
        // 记录当前子字符串的左右下标值
        int left = 0;
        int right = 0;
        // 记录当前最小子字符串的大小以及第一最后字符的下标值
        int min = Integer.MAX_VALUE;
        int minLeft = 0;
        int minRight = 0;
        for (; right < s.length(); right++) {
            char temp = s.charAt(right);
            if (map.containsKey(temp)) {// 向后遍历出所包含t的字符串
                count = map.get(temp) > 0 ? count + 1 : count;
                map.put(temp, map.get(temp) - 1);
            }
            while (count == t.length()) {// 得出一个符合条件的子字符串
                if (right - left < min) {// 更新min minLeft minRight 信息
                    min = right - left;
                    minLeft = left;
                    minRight = right;
                }
                char c = s.charAt(left);
                if (map.containsKey(c)) {// 向左收缩 判断所删减的字符是否在map中
                    if (map.get(c) >= 0)
                        count--;// count--时需要判断一下是否需要--
                    map.put(c, map.get(c) + 1);
                }
                left++;
            }
        }
        return min == Integer.MAX_VALUE ? "" : s.substring(minLeft, minRight + 1);
    }
}